### ISSUE TITLE
empty operation id with 2.1-beta4

### ISSUE THREAD

**Issue by zivkan** [Microsoft/ApplicationInsights-aspnetcore#444]
Hello,

TL;DR:  With the 2.1.0-beta4 version of this package, telemetryClient.Context.Operation.Id is null when viewed inside a MVC controller, but is not null with version 2.0.0 of the package.

Long version:  My team has an application in Service Fabric, where the ASP.NET Core app is just a reverse proxy to all the stateless services. We created a class named RequestContext where we store telementryClient's Context.Operation.Id, so that any telemetry we send from the stateless services are matched to the HTTP request. Using v2.0.0 of the package, I noticed the request time and duration was wrong, but it's a known issue fixed in the 2.1 betas (I can verify the time and duration look correct to me now). However, all of the properties in Context.Operation are null, so there's no way for me to link telemetry sent from the backend services to the http request. I tried setting Context.Operation.Id to Guid.NewGuid().ToString(), but the value gets overwritten in the ASP.NET Core app, so the dependencies still aren't linked to request.

thanks.

**Comment by yantang-msft** [comment]
Hi @zivkan , do you store the operation id of the telemetryClient or a specific telemetry?
We changed the logic of how operation id is set on a telemetry. But if you're talking about the telemetryClient, you can open an issue under this repo https://github.com/Microsoft/ApplicationInsights-dotnet

**Comment by zivkan** [comment]
Hi @yantang-msft,  I don't fully understand the question, or I don't know how to answer it.

In my MVC controller, I have TelemetryClient as a constructor parameter, so DI gives me a scoped TelemetryClient. With v2.0.0 of the nuget package, the TelemetryClient is pre-populated with an Operation.ID, that I can pass to my microservices. So, I'm getting the Operation ID from the telemtryClient, but it's a specific telemtry. Each HTTP request gets a different Operation ID. Using the v2.1.0-beta4 nuget package (with the beta versions of the AI-dotnet nuget), all properties in the Context.Operation object are null. So, there's nothing for me to pass to our microservices to link telemetry between microservices.

Which repo is the best one to report this issue? Still ApplicationInsights-dotnet?

**Comment by yantang-msft** [comment]
When I'm saying specific telemetry, I mean instance of RequestTelemetry, DependencyTelemtry...
I don't fully understand how you are trying to use the operation id of TelemetryClient to link telemetry between microservices. Seems like you want to set the operation id from TelemetryClient to some property of the telemetry you're going to send?
With latest sdk, such correlation should happen automatically, checkout this video for more details: https://channel9.msdn.com/events/Build/2017/P4113

I think we don't have any changes regarding TelemetryClient for this repo. So if you want to figure out why the TelemtryClient.Context.Operation is null, you can try opening an issue under ApplicationInsights-dotnet.

**Comment by zivkan** [comment]
Thanks for taking the time to reply to me. I'll repost my question in the other repo.

In case it helps better understand my issue, our microservices are not running ASP.NET. Similar to how the javascript App Insights client puts something in a HTTP header in XHR requests, which some .NET code (your nuget packet) uses to populate fields in a RequestTelemetry object, we need to pass some information from our ASP.NET Core app to our microservices, so that they can link telemetry together. I don't know if your nuget package will automatically add the same tracking id in outgoing http requests, but in any case our microservices are not running on HTTP, so we must manually extract the ID to pass to the microservices. In our microservices we new up our own RequestTelemetry object, and call TrackDependency, TrackTrace, TrackEvent, and so on.

Maybe the way we've been doing it so far is "wrong", but I can't find any doco on how to track a single request across multiple micro-services. Before the v2.1.0-beta4 nuget package, what we were doing was working for us. The automatic dependency tracking is awsome, but unfortunately doesn't seem to work for us.

**Comment by zivkan** [comment]
I just tested using v2.0.0 of AI-aspnetcore with AI-dotnet v2.4.0-beta3, and the operation id is filled in my MVC controller. So, the change in behaviour that I'm seeing appears to be a change in AI-aspnetcore, not AI-dotnet.

**Comment by yantang-msft** [comment]
Hi @zivkan , I just gone through the history and found it's actually a bug that you're relying on, and this has been fixed at here https://github.com/Microsoft/ApplicationInsights-aspnetcore/pull/343/files#diff-127c7a3a6227c377b206b4904744707cR42. 
You may need to find another way to link the telemetry. If your services are communicating through Http, we will add such correlation information as the Request-Id http header for outgoing request.

**Comment by yantang-msft** [comment]
Synced with @zivkan offline. The request telemetry's operation id is now set by System.Diagnostics.Activity.Current.RootId.
So just replace TelemetryClient.Context.Operation.Id by System.Diagnostics.Activity.Current.RootId will make it work again.

**Comment by zivkan** [comment]
I can confirm this works for me. Thank you very much!
